cEP-0026.md

```diff
@@ -0,0 +1,189 @@
+
+The implementation will be divided into four parts
+
+- Prototype for `FileFactory` class implementation.
+- Prototype for `Directory` class implementation.
+- Ignore directories functionality.
+- Cache control flags.
+
+### Prototype of FileFactory class implementation
+
+This class will follow the Factory design pattern and will act as a proxy to
+dealing with files. This class will have various methods to provide the file
+contents in different forms like string, binary format, etc. With the help of
+this class that we can replace the actual file contents in our file-dict with
+these objects which will itself be a performance improvement.
+
+The following is the prototype for `FileFactory` class.
+
+```python
+import linecache
+import os
+
+from coalib.core.PersistentHash import persistent_hash
+
+
+class FileFactory:
+    def __init__(self, filename, parent):
+        self.filename = filename
+        self.parent = os.path.abspath(os.path.join(self.filename, os.pardir))
+
+    @classmethod
+    def open(cls, filename, newline=None):
+        return cls(filename, newline=newline)
+
+    def file_as_bytes(self):
+        with open(self.filename, 'rb') as fp:
+            return fp.read()
+
+    def file_as_string(self):
+        return self.file_as_bytes().decode()
+
+    def file_as_list(self):
+        self.lines = linecache.getlines(self.filename)
+        self.linecount = len(self.lines)
+
+    def get_line(self, line):
+        return linecache.getline(self.filename, line)
+
+    @property
+    def timestamp(self):
+        # get the last modified timestamp of the file
+        return os.path.gettime(self.filename)
+
+    def hash(self):
+        # hash the absolute file path to get a unique hash
+        return persistent_hash(self.filename)
+```
+
+### Prototype of Directory class implementation
+
+The objects of this class will also reside in the file-set (formerly known as
+the file-dict) The objects of this class can also be used by `DirectoryBears` as
+discussed on this thread [#4999](https://github.com/coala/coala/issues/4999) by
+performing a simple type check on the entries in the dict, whether the objects
+are instance of `FileFactory` or `Directory`.
+
+Following is the prototype for the `Directory` class.
+
+```python
+import os
+
+from coalib.parsing.Globbing import glob
+
+
+class Directory:
+    def __init__(self, path):
+        self.path = path
+        self.parent = os.path.abspath(os.path.join(p, os.pardir)) # parent dir
+        self.files = 0 # number of files contained in a directory
+        self.subdirs = 0 # number of sub-directories contained in a directory
+
+    def count_children(self):
+        children = os.listdir(self.path)
+
+        for child in children:
+            if os.path.isfile(child):
+                self.files += 1
+
+        self.subdirs = len(children) - self.files
+
+    def get_dir(pattern):
+        return glob(pattern) # get sub directory matching the given glob pattern
```

Comment:
I guess what's definitely useful is a list of (maybe not ignored) file names. The `NamingConventionBear` would benefit from this.

Reply:
@Makman2 I'm sorry but I don't understand. Isn't naming convention bear intended to be used to find out unconventional identifiers in a file? Why do we need a functionality like that in the Directory class?
I think that bear is supposed on operate on files rather than dirs.

Reply:
oh sorry, I meant the `FilenameBear` :3

Reply:
But you can add this in the next iteration